### PR TITLE
[receiver/mongodbreceiver] Enhance partial error handling

### PR DESCRIPTION
**Pull request by JonathanWamsley**
**Description:**
Enhances partial error handling which includes the metric & attribute(s) in the partial error message.

**Link to tracking Issue:** 
#13367

**Testing:**
- test for partial error handling when client fails to return a payload
- test for partial error handling when metrics fail to be collected

### CODE REVIEW

File: receiver/mongodbreceiver/metrics.go [hunk 1]
```diff
@@ -43,176 +43,154 @@ var documentMap = map[string]metadata.AttributeOperation{
...
-		return
-	}
-	collectionsVal, err := parseInt(collections)
+	metricPath := []string{"collections"}
+	metricName := "mongodb.collection.count"
+	val, err := collectMetric(doc, metricPath)
 	if err != nil {
-		errs.AddPartial(1, err)
+		errs.AddPartial(1, fmt.Errorf(collectMetricWithAttributes, metricName, dbName, err))
+		s.logger.Warn(fmt.Sprintf(collectMetricWarningWithAttributes, metricName, dbName), zap.Error(err))
```
Comment (djaglowski) [inline receiver/mongodbreceiver/metrics.go:60]: I don't think we need both the error and the log

File: receiver/mongodbreceiver/metrics.go [hunk 2]
```diff
@@ -476,6 +437,18 @@ func digForCollectionPathNames(document bson.M) ([]string, error) {
...
+func collectMetric(document bson.M, path []string) (int64, error) {
+	metric, err := dig(document, path)
+	if err != nil {
+		return 0, err
+	}
+	metricVal, err := parseInt(metric)
+	if err != nil {
+		return 0, err
+	}
+	return metricVal, nil
```
Comment (djaglowski) [inline receiver/mongodbreceiver/metrics.go:449]:
```suggestion
	return parseInt(metric)
```

File: receiver/mongodbreceiver/metrics.go [hunk 3]
```diff
@@ -280,161 +250,152 @@ func (s *mongodbScraper) recordCacheOperations(now pcommon.Timestamp, doc bson.M
...
 			indexAccessValue, err := parseInt(indexAccess)
 			if err != nil {
-				errs.AddPartial(1, err)
+				errs.AddPartial(1, fmt.Errorf(collectMetricWithAttributes, metricName, metricAttributes, err))
+				s.logger.Warn(fmt.Sprintf(collectMetricWarningWithAttributes, metricName, metricAttributes), zap.Error(err))
 				return
 			}
 			indexAccessTotal += indexAccessValue
 		}
+		if len(documents) == 0 {
```
Comment (djaglowski) [inline receiver/mongodbreceiver/metrics.go:365]: Is this necessarily an error? Can a mongo instance just have no documents?

Comment (djaglowski) [inline receiver/mongodbreceiver/metrics.go:365]: If this needs to be an error, move this check before the loop.

File: unreleased/mongodbreceiver-add-partial-error-handeling.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: mongodbreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Adds partial error handling with descriptive metric/attribute messages.
```
Comment (djaglowski) [inline unreleased/mongodbreceiver-add-partial-error-handeling.yaml:8]: I think we already had partial error handling, but this PR enhances the error messages.

File: receiver/mongodbreceiver/metrics.go (hunk 1 above)
Comment (JonathanWamsley) [reply receiver/mongodbreceiver/metrics.go:60]: Would this apply in all situations, both in scraper.go calling a client to get a payload response and for adding a metric in metrics.go? I am guessing I'll keep partial error in metrics.go file and remove the log warning. What about the in scraper.go, should those partial errors stay or should the log warn stay?

File: receiver/mongodbreceiver/metrics.go (hunk 3 above)
Comment (JonathanWamsley) [reply receiver/mongodbreceiver/metrics.go]: Mongodb can have no documents testing on a fresh instance. I think a zero value is reasonable though so this check can be removed. Good call out.

Comment (JonathanWamsley) [reply receiver/mongodbreceiver/metrics.go]: not needed check

File: unreleased/mongodbreceiver-add-partial-error-handeling.yaml (hunk 4 above)
Comment (JonathanWamsley) [reply unreleased/mongodbreceiver-add-partial-error-handeling.yaml]: Updated word to enhance

**Review by djaglowski — APPROVED** [review]

**Pull request merged**